javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/resolution/naming/NameLogic.java

```diff
@@ -240,8 +276,683 @@ public static NameRole classifyRole(Node name) {
+        //
+        //      * Otherwise, this AmbiguousName is reclassified as a PackageName. A later step determines whether or not
+        //        a package of that name actually exists.
+
+        if (leftNameCategory == NameCategory.PACKAGE_NAME) {
+            if (typeSolver.hasType(nameAsString(nameNode))) {
+                return NameCategory.TYPE_NAME;
+            } else {
+                return NameCategory.PACKAGE_NAME;
+            }
+        }
+
+        // * If the name to the left of the "." is reclassified as a TypeName, then:
+        //
+        //      * If the Identifier is the name of a method or field of the type denoted by TypeName, then this
+        //        AmbiguousName is reclassified as an ExpressionName.
+        //
+        //      * Otherwise, if the Identifier is a valid TypeIdentifier and is the name of a member type of the type
+        //        denoted by TypeName, then this AmbiguousName is reclassified as a TypeName.
+        //
+        //      * Otherwise, a compile-time error occurs.
+
+        if (leftNameCategory == NameCategory.TYPE_NAME) {
+            SymbolReference<ResolvedTypeDeclaration> scopeTypeRef = JavaParserFactory.getContext(leftName, typeSolver)
+                    .solveType(NameLogic.nameAsString(leftName), typeSolver);
+            if (scopeTypeRef.isSolved()) {
+                ResolvedTypeDeclaration scopeType = scopeTypeRef.getCorrespondingDeclaration();
+                if (scopeType instanceof ResolvedReferenceTypeDeclaration) {
+                    ResolvedReferenceTypeDeclaration scopeRefType = scopeType.asReferenceType();
+                    if (scopeRefType.getAllMethods().stream().anyMatch(m -> m.getName().equals(rightName))) {
+                        return NameCategory.EXPRESSION_NAME;
+                    }
+                    if (scopeRefType.getAllFields().stream().anyMatch(f -> f.isStatic() && f.getName().equals(rightName))) {
+                        return NameCategory.EXPRESSION_NAME;
+                    }
+                    if (scopeRefType.hasInternalType(rightName)) {
+                        return NameCategory.TYPE_NAME;
+                    }
+                    return NameCategory.COMPILATION_ERROR;
+                } else {
+                    throw new UnsupportedOperationException();
+                }
+            } else {
+                throw new UnsolvedSymbolException("Unable to solve context type: " + NameLogic.nameAsString(leftName));
+            }
+        }
+
+        // * If the name to the left of the "." is reclassified as an ExpressionName, then this AmbiguousName is
+        //   reclassified as an ExpressionName. A later step determines whether or not a member with the name Identifier
+        //   actually exists.
+
+        if (leftNameCategory == NameCategory.EXPRESSION_NAME) {
+            return NameCategory.EXPRESSION_NAME;
+        }
+
+        throw new UnsupportedOperationException();
```

Comment:
What is left unsupported? Could you specify why this exception is thrown?

Reply:
All cases should be covered, however if the language evolves and there are new usages of names then this exception could be thrown

Reply:
The message of the exception should describe that reason in my opinion. Now the user will receive a random exception